tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry_test.py

```diff
@@ -0,0 +1,148 @@
+    self.cluster_preserve_quantize_registry = \
+      cluster_preserve_quantize_registry.ClusterPreserveQuantizeRegistry()
+    # layers which are supported
+    # initial and build a Conv2D layer
+    self.layer_conv2d = layers.Conv2D(10, (2, 2))
+    self.layer_conv2d.build((2, 2))
+    # initial and build a Dense layer
+    self.layer_dense = layers.Dense(10)
+    self.layer_dense.build((2, 2))
+    # initial and build a ReLU layer
+    self.layer_relu = layers.ReLU()
+    self.layer_relu.build((2, 2))
+
+    # a layer which is not supported
+    # initial and build a Custom layer
+    self.layer_custom = self.CustomLayer()
+    self.layer_custom.build()
+
+
+  class CustomLayer(layers.Layer):
+    # a simple custom layer with training weights
+    def __init__(self):
+      super(ClusterPreserveQuantizeRegistryTest.CustomLayer, self).__init__()
+
+    def build(self, input_shape=(2, 2)):
+      self.add_weight(shape=input_shape,
+                      initializer="random_normal",
+                      trainable=True)
+
+  class CustomQuantizeConfig(QuantizeConfig):
+    def get_weights_and_quantizers(self, layer):
+      return []
+
+    def get_activations_and_quantizers(self, layer):
+      return []
+
+    def set_quantize_weights(self, layer, quantize_weights):
+      pass
+
+    def set_quantize_activations(self, layer, quantize_activations):
+      pass
+
+    def get_output_quantizers(self, layer):
+      return []
+
+    def get_config(self):
+      return {}
+
+  def testSupportsKerasLayer(self):
+    # test registered layer
+    self.assertTrue(
+        self.cluster_preserve_quantize_registry.supports(self.layer_dense))
+    self.assertTrue(
+        self.cluster_preserve_quantize_registry.supports(self.layer_conv2d))
+    # test layer without training weights
+    self.assertTrue(
+        self.cluster_preserve_quantize_registry.supports(self.layer_relu))
+
+  def testDoesNotSupportCustomLayer(self):
+    self.assertFalse(
+        self.cluster_preserve_quantize_registry.supports(self.layer_custom))
+
+  def testApplyClusterPreserveWithQuantizeConfig(self):
+    self.cluster_preserve_quantize_registry.\
+      apply_cluster_preserve_quantize_config(
+        self.layer_conv2d,
+        default_8bit_quantize_registry.Default8BitConvQuantizeConfig(
+            ['kernel'], ['activation'], False))
+
+  def testRaisesErrorUnsupportedQuantizeConfigWithLayer(self):
+    with self.assertRaises(
+        ValueError, msg="Unregistered QuantizeConfigs should raise error."):
+      self.cluster_preserve_quantize_registry.\
+        apply_cluster_preserve_quantize_config(
+          self.layer_conv2d, self.CustomQuantizeConfig)
+
+    with self.assertRaises(ValueError,
+                           msg="Unregistered layers should raise error."):
+      self.cluster_preserve_quantize_registry.\
+        apply_cluster_preserve_quantize_config(
+          self.layer_custom, self.CustomQuantizeConfig)
+
```

Comment:
Ditto

Reply:
Thanks, I found similar problems in other places of the files and made changes too.